### PR TITLE
Feature #17 Add and Delete Activity Endpoints

### LINKED ISSUE #17: New Add and Delete Activity Endpoints
**Is your feature request related to a problem? Please describe.**
Currently when dragging an activity from the frontend the entire weekplan is updated. The same problem occurs when we need to add a new activity to the weekplan.

**Describe the solution you'd like**
We need to new endpoints:
1. Endpoint to delete a specific activity from a weekplan given the relevant weekplan information
2. Endpoint to add a specific activity to a weekplan given the relevant weekplan information

**Considerations**
Consider making a new AcitvityController.cs to host the new functionality
Make the endpoint v2

### PR DESCRIPTION
**Pull request by swuf**
This closes #17.

### CODE REVIEW

**Comment by Jacob6565** [comment]
@aau-giraf/sw613f19 you have been assigned this. Here's your checklist. I marked @HaZzM22 as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [ ] Is the code in the right place? (Both in terms of folder structure and class structure)
- [ ] Could this code, to the best of your knowledge, have reused existing code?
- [ ] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [ ] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [ ] Are names meaningful and self-documenting?
- [ ] Is it understandable, by reading the code, what it does?
- [ ] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [ ] Is it understandable, by reading the tests, what they do?
- [ ] Do the tests cover sensible cases? Both happy paths and exception paths?
- [ ] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [ ] Are error messages in the code understandable?
- [ ] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [ ] Are the features that the code claims to implement, actually implemented?
- [ ] Does it look like the code has bugs?
- [ ] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [ ] Are UI messages, that the user might meet, actually helpful?
- [ ] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [ ] Is there well written code in this? Remember to tell the developer!

**Comment by Jacob6565** [comment]
@aau-giraf/sw612f19 you have been assigned this. Here's your checklist. I marked @thomaslhansen as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [ ] Is the code in the right place? (Both in terms of folder structure and class structure)
- [ ] Could this code, to the best of your knowledge, have reused existing code?
- [ ] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [ ] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [ ] Are names meaningful and self-documenting?
- [ ] Is it understandable, by reading the code, what it does?
- [ ] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [ ] Is it understandable, by reading the tests, what they do?
- [ ] Do the tests cover sensible cases? Both happy paths and exception paths?
- [ ] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [ ] Are error messages in the code understandable?
- [ ] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [ ] Are the features that the code claims to implement, actually implemented?
- [ ] Does it look like the code has bugs?
- [ ] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [ ] Are UI messages, that the user might meet, actually helpful?
- [ ] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [ ] Is there well written code in this? Remember to tell the developer!

File: GirafRest/Controllers/ActivityController.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+            // check access rights
+            if (!(await _authentication.HasEditOrReadUserAccess(await _giraf._userManager.GetUserAsync(HttpContext.User), user)))
+                return new ErrorResponse(errorCode: ErrorCode.NotAuthorized);
+
+            // throws error if none of user's weeks' has the specific activity
+            if (!user.WeekSchedule.Any(w => w.Weekdays.Any(wd => wd.Activities.Any(act => act.Key == activityId))))
+                return new ErrorResponse(errorCode: ErrorCode.ActivityNotFound);
+
+            List<Activity> a = _giraf._context.Activities.ToList(); ;
```
Comment (ralled96) [inline GirafRest/Controllers/ActivityController.cs]: double ;

File: GirafRest/Controllers/ActivityController.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        [Authorize]
+        public async Task<Response> DeleteActivity(string userId, long activityId)
+        {
+            GirafUser user = await _giraf.LoadUserWithWeekSchedules(userId);
+            if (user == null)
+                return new ErrorResponse(ErrorCode.UserNotFound);
+
+            // check access rights
+            if (!(await _authentication.HasEditOrReadUserAccess(await _giraf._userManager.GetUserAsync(HttpContext.User), user)))
+                return new ErrorResponse(errorCode: ErrorCode.NotAuthorized);
```
Comment (ralled96) [inline GirafRest/Controllers/ActivityController.cs]: consistency with ErrorResponse, line 88, 92 and 96? 😃

File: GirafRest/Controllers/ActivityController.cs (hunk 1 above)
Comment (ralled96) [inline GirafRest/Controllers/ActivityController.cs]: variable a is never used?

File: GirafRest/Controllers/ActivityController.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            List<Activity> a = _giraf._context.Activities.ToList(); ;
+            Activity targetActivity = _giraf._context.Activities.First(act => act.Key == activityId);
+
+            _giraf._context.Activities.Remove(targetActivity);
+            await _giraf._context.SaveChangesAsync();
+
+            return new Response();
+        }
+
+        private bool ActivityExists(long id)
```
Comment (ralled96) [inline GirafRest/Controllers/ActivityController.cs]: This method is never used

File: GirafRest/Controllers/ActivityController.cs (hunk 1 above)
Comment (swuf) [reply GirafRest/Controllers/ActivityController.cs]: Removed (:

Comment (swuf) [reply GirafRest/Controllers/ActivityController.cs]: Removed :)

File: GirafRest/Controllers/ActivityController.cs (hunk 3 above)
Comment (swuf) [reply GirafRest/Controllers/ActivityController.cs]: Removed (:

File: GirafRest/Controllers/ActivityController.cs (hunk 2 above)
Comment (swuf) [reply GirafRest/Controllers/ActivityController.cs]: ErrorResponse syntax is now consistent 🐤

**Comment by KVISDAOWNER** [comment]
We intend to add integration tests before merging

**Comment by MStenkaer** [comment]
Why is the ErrorCode 'NotFound' changed to 'ActivityNotFound'?

**Comment by KVISDAOWNER** [comment]
> Why is the ErrorCode 'NotFound' changed to 'ActivityNotFound'?

Initially it was changed to be more sayinng and follow the convention of e.g. 'UserNotFound' and 'WeekNotFound'. But some of the 'NotFound' errorcodes that do not regard an activity have been changed to 'ActivityNotFound'. This will be fixed before merging.

**Pull request merged**